File: src/transcriber.ts
/**
 * Transcription and subtitle parsing for YouTube videos.
 */

import { readFile } from 'fs/promises';

// Types
export interface Segment {
  text: string;
  start_time: number;
  end_time: number;
}

/**
 * Parse VTT timestamp to seconds.
 * Supports formats:
 * - 00:00:00.000 (hours:minutes:seconds.milliseconds)
 * - 00:00.000 (minutes:seconds.milliseconds)
 */
function parseVttTimestamp(timestamp: string): number {
  const parts = timestamp.trim().split(':');
  let hours = 0;
  let minutes: number;
  let seconds: number;

  if (parts.length === 3) {
    hours = parseInt(parts[0], 10);
    minutes = parseInt(parts[1], 10);
    seconds = parseFloat(parts[2]);
  } else if (parts.length === 2) {
    minutes = parseInt(parts[0], 10);
    seconds = parseFloat(parts[1]);
  } else {
    throw new Error(`Invalid VTT timestamp format: ${timestamp}`);
  }

  return hours * 3600 + minutes * 60 + seconds;
}

/**
 * Parse SRT timestamp to seconds.
 * Format: 00:00:00,000 (hours:minutes:seconds,milliseconds)
 */
function parseSrtTimestamp(timestamp: string): number {
  // SRT uses comma as decimal separator
  const normalized = timestamp.trim().replace(',', '.');
  const parts = normalized.split(':');

  if (parts.length !== 3) {
    throw new Error(`Invalid SRT timestamp format: ${timestamp}`);
  }

  const hours = parseInt(parts[0], 10);
  const minutes = parseInt(parts[1], 10);
  const seconds = parseFloat(parts[2]);

  return hours * 3600 + minutes * 60 + seconds;
}

/**
 * Deduplicate VTT segments that have rolling/scrolling text.
 * YouTube captions often show scrolling text where each cue contains
 * previous text plus new text.
 */
function deduplicateVttSegments(segments: Segment[]): Segment[] {
  if (segments.length === 0) return [];

  const result: Segment[] = [];
  let prevText = '';

  for (const segment of segments) {
    const text = segment.text;
    let newText: string;

    // If current text starts with previous text, extract only new part
    if (prevText && text.startsWith(prevText)) {
      newText = text.slice(prevText.length).trim();
    } else if (prevText && text.includes(prevText)) {
      // Previous text is somewhere in current text - extract after it
      const idx = text.indexOf(prevText);
      newText = text.slice(idx + prevText.length).trim();
    } else {
      newText = text;
    }

    if (newText) {
      result.push({
        text: newText,
        start_time: segment.start_time,
        end_time: segment.end_time,
      });
    }

    prevText = text;
  }

  return result;
}

/**
 * Parse a VTT subtitle file.
 */
export async function parseVtt(filePath: string): Promise<Segment[]> {
  const content = await readFile(filePath, 'utf-8');
  const segments: Segment[] = [];

  const lines = content.split('\n');
  let startIdx = 0;

  // Find WEBVTT header
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i].trim();
    if (line === 'WEBVTT' || line.startsWith('WEBVTT ')) {
      startIdx = i + 1;
      break;
    }
  }

  // Skip header metadata until first timestamp
  while (startIdx < lines.length) {
    const line = lines[startIdx].trim();
    if (line.includes('-->')) break;
    startIdx++;
  }

  // Parse cues
  let i = startIdx;
  while (i < lines.length) {
    const line = lines[i].trim();

    // Skip empty lines and cue identifiers
    if (!line || (line && !line.includes('-->') && i + 1 < lines.length && lines[i + 1].includes('-->'))) {
      i++;
      continue;
    }

    // Look for timestamp line
    if (line.includes('-->')) {
      const timestampMatch = line.match(/([\d:.]+)\s*-->\s*([\d:.]+)/);
      if (timestampMatch) {
        const startTime = parseVttTimestamp(timestampMatch[1]);
        const endTime = parseVttTimestamp(timestampMatch[2]);

        // Collect text lines
        const textLines: string[] = [];
        i++;
        while (i < lines.length) {
          const textLine = lines[i];
          // Stop at empty line or next timestamp
          if (!textLine.trim()) break;
          if (textLine.includes('-->')) {
            i--;
            break;
          }
          // Only keep lines with inline timestamps (new content)
          const hasInlineTimestamps = textLine.includes('<c>') || /\d{2}:\d{2}/.test(textLine);
          if (hasInlineTimestamps || textLines.length === 0) {
            // Remove VTT formatting tags
            let cleanLine = textLine.replace(/<[^>]+>/g, '');
            if (hasInlineTimestamps) {
              textLines.push(cleanLine.trim());
            }
          }
          i++;
        }

        const text = textLines.join(' ').trim();
        if (text) {
          segments.push({ text, start_time: startTime, end_time: endTime });
        }
      }
    }

    i++;
  }

  // Deduplicate rolling/scrolling text
  return deduplicateVttSegments(segments);
}

/**
 * Parse an SRT subtitle file.
 */
export async function parseSrt(filePath: string): Promise<Segment[]> {
  const content = await readFile(filePath, 'utf-8');
  const segments: Segment[] = [];

  // Split by double newlines to get blocks
  const blocks = content.trim().split(/\n\n+/);

  for (const block of blocks) {
    const lines = block.trim().split('\n');
    if (lines.length < 2) continue;

    // Find the timestamp line
    let timestampIdx: number | null = null;
    for (let idx = 0; idx < lines.length; idx++) {
      if (lines[idx].includes('-->')) {
        timestampIdx = idx;
        break;
      }
    }

    if (timestampIdx === null) continue;

    // Parse timestamp
    const timestampLine = lines[timestampIdx];
    const timestampMatch = timestampLine.match(/([\d:,]+)\s*-->\s*([\d:,]+)/);
    if (!timestampMatch) continue;

    const startTime = parseSrtTimestamp(timestampMatch[1]);
    const endTime = parseSrtTimestamp(timestampMatch[2]);

    // Text is everything after the timestamp line
    const textLines = lines.slice(timestampIdx + 1);
    let text = textLines.join(' ');
    // Remove SRT formatting tags
    text = text.replace(/<[^>]+>/g, '');
    text = text.replace(/\{[^}]+\}/g, '');
    text = text.trim();

    if (text) {
      segments.push({ text, start_time: startTime, end_time: endTime });
    }
  }

  return segments;
}

/**
 * Auto-detect subtitle format and parse.
 */
export async function parseSubtitles(filePath: string): Promise<Segment[]> {
  const suffix = filePath.toLowerCase().split('.').pop();

  if (suffix === 'vtt') {
    return parseVtt(filePath);
  } else if (suffix === 'srt') {
    return parseSrt(filePath);
  } else {
    // Try to detect from content
    const content = await readFile(filePath, 'utf-8');
    if (content.trim().startsWith('WEBVTT')) {
      return parseVtt(filePath);
    }
    // Check for SRT pattern
    if (/^\d+\s*\n\d{2}:\d{2}:\d{2},\d{3}/m.test(content)) {
      return parseSrt(filePath);
    }

    throw new Error(
      `Cannot determine subtitle format for ${filePath}. ` +
      'Expected .vtt or .srt extension, or recognizable content.'
    );
  }
}

/**
 * Transcribe audio using ElevenLabs Scribe.
 */
export async function transcribeAudio(audioPath: string): Promise<Segment[]> {
  const apiKey = process.env.ELEVENLABS_API_KEY;
  if (!apiKey) {
    throw new Error(
      'ELEVENLABS_API_KEY environment variable is not set. ' +
      'Please set it to use ElevenLabs transcription.'
    );
  }

  const { ElevenLabsClient } = await import('@elevenlabs/elevenlabs-js');
  const { createReadStream } = await import('fs');

  const client = new ElevenLabsClient({ apiKey });

  const audioFile = createReadStream(audioPath);
  const result = await client.speechToText.convert({
    file: audioFile,
    modelId: 'scribe_v1',
  });

  // Group words into sentence segments
  return groupWordsIntoSegments(result);
}

/**
 * Group word-level timestamps into sentence segments.
 */
function groupWordsIntoSegments(transcriptionResult: any): Segment[] {
  const segments: Segment[] = [];

  if (!transcriptionResult.words || transcriptionResult.words.length === 0) {
    // If no word-level data, return the full text as one segment
    if (transcriptionResult.text) {
      return [{
        text: transcriptionResult.text.trim(),
        start_time: 0,
        end_time: 0,
      }];
    }
    return [];
  }

  const words = transcriptionResult.words;
  let currentSegmentWords: string[] = [];
  let currentStart: number | null = null;

  const sentenceEndings = new Set(['.', '!', '?']);

  for (const word of words) {
    const wordText = word.text || '';
    const wordStart = word.start || 0;
    const wordEnd = word.end || 0;

    if (currentStart === null) {
      currentStart = wordStart;
    }

    currentSegmentWords.push(wordText);

    // Check if this word ends a sentence
    if (wordText && sentenceEndings.has(wordText[wordText.length - 1])) {
      const text = currentSegmentWords.join(' ').trim();
      if (text && currentStart !== null) {
        segments.push({
          text,
          start_time: currentStart,
          end_time: wordEnd,
        });
      }
      currentSegmentWords = [];
      currentStart = null;
    }
  }

  // Handle remaining words
  if (currentSegmentWords.length > 0) {
    const text = currentSegmentWords.join(' ').trim();
    if (text && currentStart !== null) {
      const lastWord = words[words.length - 1];
      segments.push({
        text,
        start_time: currentStart,
        end_time: lastWord.end ?? 0,
      });
    }
  }

  return segments;
}

/**
 * Normalize and clean up transcript segments.
 */
export function normalizeSegments(
  segments: Segment[],
  minDuration: number = 0.5,
  mergeThreshold: number = 1.0
): Segment[] {
  if (segments.length === 0) return [];

  const normalized: Segment[] = [];

  for (const segment of segments) {
    // Clean up text
    let text = segment.text;
    text = text.replace(/\s+/g, ' ').trim();

    if (!text) continue;

    let startTime = segment.start_time;
    let endTime = segment.end_time;

    // Ensure start_time < end_time
    if (startTime >= endTime) {
      if (startTime > endTime) {
        [startTime, endTime] = [endTime, startTime];
      } else {
        endTime = startTime + 0.1;
      }
    }

    normalized.push({ text, start_time: startTime, end_time: endTime });
  }

  // Merge very short segments
  if (minDuration > 0) {
    return mergeShortSegments(normalized, minDuration, mergeThreshold);
  }

  return normalized;
}

/**
 * Merge segments that are too short with adjacent segments.
 */
function mergeShortSegments(
  segments: Segment[],
  minDuration: number,
  mergeThreshold: number
): Segment[] {
  if (segments.length === 0) return [];

  const result: Segment[] = [];
  let current = { ...segments[0] };

  for (let i = 1; i < segments.length; i++) {
    const nextSeg = segments[i];
    const currentDuration = current.end_time - current.start_time;
    const gap = nextSeg.start_time - current.end_time;

    if (currentDuration < minDuration && gap <= mergeThreshold) {
      // Merge with next segment
      current.text = current.text + ' ' + nextSeg.text;
      current.end_time = nextSeg.end_time;
    } else {
      result.push(current);
      current = { ...nextSeg };
    }
  }

  result.push(current);
  return result;
}

File: channel_chat/transcriber.py
"""Transcription and subtitle parsing for YouTube videos."""

import os
import re
from pathlib import Path


def _parse_vtt_timestamp(timestamp: str) -> float:
    """Parse VTT timestamp to seconds.

    Supports formats:
    - 00:00:00.000 (hours:minutes:seconds.milliseconds)
    - 00:00.000 (minutes:seconds.milliseconds)
    """
    parts = timestamp.strip().split(":")
    if len(parts) == 3:
        hours, minutes, seconds = parts
        hours = int(hours)
    elif len(parts) == 2:
        hours = 0
        minutes, seconds = parts
    else:
        raise ValueError(f"Invalid VTT timestamp format: {timestamp}")

    minutes = int(minutes)
    seconds = float(seconds)

    return hours * 3600 + minutes * 60 + seconds


def _parse_srt_timestamp(timestamp: str) -> float:
    """Parse SRT timestamp to seconds.

    Format: 00:00:00,000 (hours:minutes:seconds,milliseconds)
    """
    # SRT uses comma as decimal separator
    timestamp = timestamp.strip().replace(",", ".")
    parts = timestamp.split(":")

    if len(parts) != 3:
        raise ValueError(f"Invalid SRT timestamp format: {timestamp}")

    hours = int(parts[0])
    minutes = int(parts[1])
    seconds = float(parts[2])

    return hours * 3600 + minutes * 60 + seconds


def parse_vtt(file_path: Path) -> list[dict]:
    """Parse a VTT subtitle file.

    Args:
        file_path: Path to the VTT file.

    Returns:
        List of segment dicts with {text, start_time, end_time}.
    """
    content = file_path.read_text(encoding="utf-8")
    segments = []

    # Remove WEBVTT header and any metadata
    lines = content.split("\n")
    start_idx = 0
    for i, line in enumerate(lines):
        if line.strip() == "WEBVTT" or line.strip().startswith("WEBVTT "):
            start_idx = i + 1
            break

    # Skip any header metadata (lines before first timestamp)
    while start_idx < len(lines):
        line = lines[start_idx].strip()
        if "-->" in line:
            break
        start_idx += 1

    # Parse cues
    i = start_idx
    while i < len(lines):
        line = lines[i].strip()

        # Skip empty lines and cue identifiers
        if not line or (line and "-->" not in line and not lines[i - 1].strip() == ""):
            # Check if this might be a cue identifier (numeric or string before timestamp)
            if i + 1 < len(lines) and "-->" in lines[i + 1]:
                i += 1
                continue

        # Look for timestamp line
        if "-->" in line:
            # Parse timestamp line: "00:00:00.000 --> 00:00:05.000" with optional settings
            timestamp_match = re.match(
                r"([\d:.]+)\s*-->\s*([\d:.]+)",
                line
            )
            if timestamp_match:
                start_time = _parse_vtt_timestamp(timestamp_match.group(1))
                end_time = _parse_vtt_timestamp(timestamp_match.group(2))

                # Collect text lines until empty line or next timestamp
                text_lines = []
                i += 1
                while i < len(lines):
                    text_line = lines[i]
                    # Stop at empty line or next timestamp
                    if not text_line.strip():
                        break
                    if "-->" in text_line:
                        i -= 1  # Back up so outer loop catches this timestamp
                        break
                    # Remove VTT formatting tags like <c>, </c>, <00:00:00.000>
                    clean_line = re.sub(r"<[^>]+>", "", text_line)
                    text_lines.append(clean_line.strip())
                    i += 1

                text = " ".join(text_lines).strip()
                if text:
                    segments.append({
                        "text": text,
                        "start_time": start_time,
                        "end_time": end_time,
                    })

        i += 1

    return segments


def parse_srt(file_path: Path) -> list[dict]:
    """Parse an SRT subtitle file.

    Args:
        file_path: Path to the SRT file.

    Returns:
        List of segment dicts with {text, start_time, end_time}.
    """
    content = file_path.read_text(encoding="utf-8")
    segments = []

    # SRT format: sequence number, timestamp line, text, blank line
    # Split by double newlines to get blocks
    blocks = re.split(r"\n\n+", content.strip())

    for block in blocks:
        lines = block.strip().split("\n")
        if len(lines) < 2:
            continue

        # Find the timestamp line (contains "-->")
        timestamp_idx = None
        for idx, line in enumerate(lines):
            if "-->" in line:
                timestamp_idx = idx
                break

        if timestamp_idx is None:
            continue

        # Parse timestamp
        timestamp_line = lines[timestamp_idx]
        timestamp_match = re.match(
            r"([\d:,]+)\s*-->\s*([\d:,]+)",
            timestamp_line
        )
        if not timestamp_match:
            continue

        start_time = _parse_srt_timestamp(timestamp_match.group(1))
        end_time = _parse_srt_timestamp(timestamp_match.group(2))

        # Text is everything after the timestamp line
        text_lines = lines[timestamp_idx + 1:]
        # Remove SRT formatting tags like <i>, </i>, {\\an8}, etc.
        text = " ".join(text_lines)
        text = re.sub(r"<[^>]+>", "", text)
        text = re.sub(r"\{[^}]+\}", "", text)
        text = text.strip()

        if text:
            segments.append({
                "text": text,
                "start_time": start_time,
                "end_time": end_time,
            })

    return segments


def parse_subtitles(file_path: Path) -> list[dict]:
    """Auto-detect subtitle format and parse.

    Args:
        file_path: Path to the subtitle file (.vtt or .srt).

    Returns:
        List of segment dicts with {text, start_time, end_time}.

    Raises:
        ValueError: If the file format cannot be determined.
    """
    file_path = Path(file_path)
    suffix = file_path.suffix.lower()

    if suffix == ".vtt":
        return parse_vtt(file_path)
    elif suffix == ".srt":
        return parse_srt(file_path)
    else:
        # Try to detect from content
        content = file_path.read_text(encoding="utf-8")
        if content.strip().startswith("WEBVTT"):
            return parse_vtt(file_path)
        # Check for SRT pattern (number followed by timestamp with comma)
        if re.search(r"^\d+\s*\n\d{2}:\d{2}:\d{2},\d{3}", content, re.MULTILINE):
            return parse_srt(file_path)

        raise ValueError(
            f"Cannot determine subtitle format for {file_path}. "
            "Expected .vtt or .srt extension, or recognizable content."
        )


def transcribe_audio(audio_path: Path) -> list[dict]:
    """Transcribe audio using ElevenLabs Scribe v2.

    Args:
        audio_path: Path to the audio file.

    Returns:
        List of segment dicts with {text, start_time, end_time}.
        Words are grouped into sentence segments.

    Raises:
        ValueError: If ELEVENLABS_API_KEY environment variable is not set.
        RuntimeError: If transcription fails.
    """
    api_key = os.environ.get("ELEVENLABS_API_KEY")
    if not api_key:
        raise ValueError(
            "ELEVENLABS_API_KEY environment variable is not set. "
            "Please set it to use ElevenLabs transcription."
        )

    from elevenlabs import ElevenLabs

    client = ElevenLabs(api_key=api_key)

    audio_path = Path(audio_path)
    with open(audio_path, "rb") as audio_file:
        # Use Speech to Text API (Scribe v2)
        result = client.speech_to_text.convert(
            file=audio_file,
            model_id="scribe_v1",  # Scribe v1 is the current model name in the API
        )

    # Group words into sentence segments
    segments = _group_words_into_segments(result)

    return segments


def _group_words_into_segments(transcription_result) -> list[dict]:
    """Group word-level timestamps into sentence segments.

    Args:
        transcription_result: ElevenLabs transcription result object.

    Returns:
        List of segment dicts with {text, start_time, end_time}.
    """
    segments = []

    # ElevenLabs returns words with timestamps
    # We need to group them into logical segments (sentences)
    if not hasattr(transcription_result, "words") or not transcription_result.words:
        # If no word-level data, return the full text as one segment
        if hasattr(transcription_result, "text") and transcription_result.text:
            return [{
                "text": transcription_result.text.strip(),
                "start_time": 0.0,
                "end_time": 0.0,
            }]
        return []

    words = transcription_result.words
    current_segment_words = []
    current_start = None

    # Sentence-ending punctuation
    sentence_endings = {".", "!", "?"}

    for word in words:
        word_text = word.text if hasattr(word, "text") else str(word.get("text", ""))
        word_start = word.start if hasattr(word, "start") else word.get("start", 0)
        word_end = word.end if hasattr(word, "end") else word.get("end", 0)

        if current_start is None:
            current_start = word_start

        current_segment_words.append(word_text)

        # Check if this word ends a sentence
        if word_text and word_text[-1] in sentence_endings:
            text = " ".join(current_segment_words).strip()
            if text:
                segments.append({
                    "text": text,
                    "start_time": current_start,
                    "end_time": word_end,
                })
            current_segment_words = []
            current_start = None

    # Handle remaining words (incomplete sentence at the end)
    if current_segment_words:
        text = " ".join(current_segment_words).strip()
        if text:
            last_word = words[-1]
            end_time = last_word.end if hasattr(last_word, "end") else last_word.get("end", 0)
            segments.append({
                "text": text,
                "start_time": current_start,
                "end_time": end_time,
            })

    return segments


def normalize_segments(
    segments: list[dict],
    min_duration: float = 0.5,
    merge_threshold: float = 1.0,
) -> list[dict]:
    """Normalize and clean up transcript segments.

    Args:
        segments: List of segment dicts with {text, start_time, end_time}.
        min_duration: Minimum segment duration in seconds. Shorter segments
            will be merged with adjacent ones.
        merge_threshold: Maximum gap (in seconds) between segments to merge
            when a segment is too short.

    Returns:
        Normalized list of segment dicts.
    """
    if not segments:
        return []

    normalized = []

    for segment in segments:
        # Clean up text
        text = segment["text"]
        # Normalize whitespace
        text = re.sub(r"\s+", " ", text).strip()
        # Remove leading/trailing punctuation-only content
        text = text.strip()

        if not text:
            continue

        start_time = float(segment["start_time"])
        end_time = float(segment["end_time"])

        # Ensure start_time < end_time
        if start_time >= end_time:
            # Try to fix by swapping or using a minimal duration
            if start_time > end_time:
                start_time, end_time = end_time, start_time
            else:
                # start_time == end_time, add minimal duration
                end_time = start_time + 0.1

        normalized.append({
            "text": text,
            "start_time": start_time,
            "end_time": end_time,
        })

    # Merge very short segments
    if min_duration > 0:
        normalized = _merge_short_segments(normalized, min_duration, merge_threshold)

    return normalized


def _merge_short_segments(
    segments: list[dict],
    min_duration: float,
    merge_threshold: float,
) -> list[dict]:
    """Merge segments that are too short with adjacent segments.

    Args:
        segments: List of normalized segments.
        min_duration: Minimum segment duration.
        merge_threshold: Maximum gap to allow merging.

    Returns:
        List of segments with short ones merged.
    """
    if not segments:
        return []

    result = []
    current = segments[0].copy()

    for i in range(1, len(segments)):
        next_seg = segments[i]
        current_duration = current["end_time"] - current["start_time"]

        # Check if current segment is too short and can be merged
        gap = next_seg["start_time"] - current["end_time"]

        if current_duration < min_duration and gap <= merge_threshold:
            # Merge with next segment
            current["text"] = current["text"] + " " + next_seg["text"]
            current["end_time"] = next_seg["end_time"]
        else:
            result.append(current)
            current = next_seg.copy()

    # Don't forget the last segment
    result.append(current)

    return result
